### PR TITLE
New `Text` type supporting O(1) size, O(log n) indexing, take, drop, append

### PR DESCRIPTION
**Pull request by pchiusano**
This PR introduces a new utility type `Rope` and uses it to (re)implement the `Bytes` type and a new `Text` type supporting fast append, indexing, take, and drop. It's intended to make it unnecessary to have a separate text buffer type (since appends are efficient) while also working well for general use. For core data types like `Text` I am a big fan of data structures with reasonable asymptotics for all common operations.

In the common case where the text is small (currently < 512 characters), it will be represented with a single chunk of type `Data.Text` with its size cached. Larger text values will be represented as trees of chunks (of size up to 512 characters).

The new text type is just used at runtime and this change doesn't affect the AST.

The new `Rope` type is currently a simple size-balanced binary tree. I suspect its constant factors will be  good but haven't done benchmarking. We might want to play with its representation later (like I considered a specialized finger tree for the implementation) and this PR makes that easy - the `Rope` module hides its constructors and has an explicit export list. It is also easy to swap out the chunk type used (like we might want a leaner UTF8 chunk type, or a different byte array type for `Bytes`).

The existing tests caught several bugs and I also added some unit tests which caught a few more. I feel pretty good about the test coverage.

Some kind of interesting things about the implementation:

* The `Eq` and `Ord` instances have to quotient out how the two ropes are chunked. It does this in sort of a clever way, lazily producing two aligned lists of chunks that can be compared like any other list.
* The sizes are an unpacked to `!Int` and we're smart about avoiding recomputing sizes all over the place. 
* Append of two similarly sized (within a factor of 2) trees is O(1).

### CODE REVIEW

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+{-# Language DeriveFoldable #-}
+
+module Unison.Util.Rope where
+
+import Prelude hiding (drop,take,reverse,map,traverse)
+import Data.Foldable (toList)
+
+data Rope a
+  = Empty
+  | One {-# unpack #-} !Int {-# unpack #-} !a
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Rope.hs:11]: Probably not necessary to start with, but in the future may be nice to generalize over the measure to allow tracking things like newlines 😄 

Adds some annoying boilerplate in terms of a `Measured` typeclass and would likely hurt the ability to unpack; so what you have is probably best for now.

File: parser-typechecker/src/Unison/Runtime/ANF.hs [hunk 2]
```diff
@@ -92,7 +92,8 @@ import Unison.Util.Bytes (Bytes)
 import qualified Unison.Util.Pretty as Pretty
 import qualified Data.Map as Map
 import qualified Data.Set as Set
-import qualified Data.Text as Text
+import qualified Data.Text as Data.Text
+import qualified Unison.Util.Text as Util.Text
```
Comment (pchiusano) [inline parser-typechecker/src/Unison/Runtime/ANF.hs:96]: This file needed both `Data.Text` and `Util.Text` so I imported them both qualified under a more explicit name for clarity.

**Review by pchiusano — COMMENTED** [review]
Self-review

* Updates to the runtime were very mechanical. A lot of replacing `Text.pack/unpack` with `Util.Text.pack/unpack`. I removed the `BuiltinForeign Data.Text` instance which helped catch all the places that needed updating.

File: parser-typechecker/src/Unison/Util/Bytes.hs [hunk 3]
```diff
@@ -1,42 +1,78 @@
...
+fromStorable v =
+  -- is there a more efficient implementation of this
+  V.generate (SV.length v) (SV.unsafeIndex v)
+{-# inline fromStorable #-}
+
+toStorable :: (V.Prim a, SV.Storable a) => V.Vector a -> SV.Vector a
+toStorable v =
+  -- is there a more efficient implementation of this
+  SV.generate (V.length v) (V.unsafeIndex v)
+{-# inline toStorable #-}
```
Comment (pchiusano) [inline parser-typechecker/src/Unison/Util/Bytes.hs]: @dolio can you help with figuring out the incantation to do one or both of these using a block copy of some sort? 

What's here isn't wrong but I think it can be done using what ends up being a `memcopy`.

File: parser-typechecker/src/Unison/Util/Bytes.hs [hunk 4]
```diff
@@ -1,42 +1,78 @@
...
+import qualified Data.Vector.Primitive as V
+import qualified Data.Vector.Storable as SV
+import qualified Data.Vector.Storable.ByteString as BSV
+import Foreign.Storable (pokeByteOff)
 import qualified Codec.Compression.Zlib as Zlib
 import qualified Codec.Compression.GZip as GZip
 
--- Block is just `newtype Block a = Block ByteArray#`
-type ByteString = Block Word8
+type ByteString = V.Vector Word8
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Bytes.hs]: Maybe call this `ShortByteString` or something to help imply when you would use this instead of Bytes?
Or is this not meant to be exported, in which case maybe add an export list?
If I just saw `ByteString` and `Bytes` in a file it's not clear which is used for what 😄

File: parser-typechecker/src/Unison/Util/Bytes.hs [hunk 5]
```diff
@@ -1,42 +1,78 @@
...
 
 null :: Bytes -> Bool
-null (Bytes bs) = T.null bs
+null = R.null . underlying
 
 empty :: Bytes
-empty = Bytes mempty
+empty = mempty
+
+isAscii :: Bytes -> Bool
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Bytes.hs:78]: Doesn't have to be as part of this PR of course, but it would likely be good to add time complexities to the docs for each of these so it's easy to see at a glance what performance we can expect when using them, e.g. at a glance looks like this is `O(n)`

File: parser-typechecker/src/Unison/Util/Bytes.hs (hunk 3 above)
Comment (dolio) [reply parser-typechecker/src/Unison/Util/Bytes.hs]: Done. There are tests for this stuff, right? I had to use some somewhat unsafe stuff.

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+instance Sized a => Sized (Rope a) where
+  size = \case
+    Empty -> 0
+    One a -> size a
+    Two n _ _ -> n
+
+null :: Sized a => Rope a -> Bool
+null r = size r == 0
+
+flatten :: Monoid a => Rope a -> a
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Util/Rope.hs:67]: Just FYI, this is also given by `Data.Foldable.fold`

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  if size a == 0 then r
+  else cons' (size a) a r
+
+snoc :: (Sized a, Semigroup a) => Rope a -> a -> Rope a
+snoc as a =
+  if size a == 0 then as
+  else snoc' as (size a) a
+
+cons' :: (Sized a, Semigroup a) => Int -> a -> Rope a -> Rope a
+cons' 0 _ as = as
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Util/Rope.hs:107]: Performance nit: It looks to me like `cons'` has an invariant that `a` is non-empty, and thus we could elide this extra size check. Is that right?

Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Util/Rope.hs:107]: Ditto for `snoc'`

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    Two sz One{} _ -> Two (sz+szN) as (One aN)
+    Two _ l r -> l <> go r aN
+
+instance (Sized a, Index a ch) => Index (Rope a) ch where
+  index i = \case
+    One a -> index i a
+    Two sz l r
+      | i < size l -> index i l
+      | i >= sz    -> Nothing
+      | otherwise  -> index (i - size l) r
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Util/Rope.hs]: Performance nit: here we know that `i` is in bounds so we could call out to a variant of `index` that does not need to perform the `i >= sz` check(s) again

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  (chunks, singleton, one, map, traverse, null, flatten, two, cons, uncons, snoc, unsnoc, 
+   Sized(..), Take(..), Drop(..), Reverse(..), Index(..), Rope,
+   ) 
+where
+
+import Prelude hiding (drop,take,reverse,map,traverse,null)
+import Data.Foldable (toList)
+import Control.DeepSeq (NFData(..))
+
+-- | Size-balanced binary tree of chunks.
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Util/Rope.hs]: Is "size-balanced binary tree of chunks" not a 100% accurate description due to `take`/`drop` not re-balancing?

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 8 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Util/Rope.hs]: Ditto for `take`/`drop`'s recursive calls in the `Two` case

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+import Prelude hiding (drop,take,reverse,map,traverse,null)
+import Data.Foldable (toList)
+import Control.DeepSeq (NFData(..))
+
+-- | Size-balanced binary tree of chunks. 
+-- The `Int` in the `Two` constructor is a cached size of that subtree.
+data Rope a
+  = Empty
+  | One {-# unpack #-} !a
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Util/Rope.hs:22]: Nit: unpack doesn't work on polymorphic fields

**Review by mitchellwrosen — APPROVED** [review]
Looks cool. :)

I only left a cursory review of nits, as I have no experience with this sort of thing. I think benchmarks for this work are important to have (eventually).

One question I have about the implementation: what's the significance of the 512 threshold in `Text`, but 32 threshold in `Rope`? I noticed that if you concatenate two `Text` that together exceed 32 characters, you get a `Two`, but I'm not sure if that's intentional.

And a couple of random questions that came to mind as I was scanning:

- Is there anything on hackage that is similar, perhaps some rope or finger tree package? Maybe a straw man: what about `Seq Text`?
- I wonder how immutable strings work in other languages. What's a JavaScript string? What's an OCaml string?
- Any relevant papers or blogpost/folklore that inspired these structures?

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 8 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs]: Good call.

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 10 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs:22]: The compiler accepts it if the field is strict (but not if it's lazy), which made me think a `newtype Text = Text (Rope Data.Text)` where the `a` is now specialized could unpack the `Data.Text` into the constructor. Do you know for sure that won't happen?

Or is there some other way to tell Haskell to produce specialized versions of `Rope` for particular instantiations of `a`?

cc @dolio

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+chunks :: Rope a -> [a]
+chunks = toList
+
+singleton, one :: Sized a => a -> Rope a
+one a | size a == 0 = Empty
+one a = One a
+singleton = one
+
+map :: Sized b => (a -> b) -> Rope a -> Rope b
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Rope.hs:38]: `map` and `traverse` don't seem to re-balance afterwards.
If there's an invariant that the caller of map shouldn't change the size of the elements during mapping/traversing then that should probably be indicated; especially since a type-changing update could very easily do so.

It'd also be worth noting that the `a`'s that are being mapped will be chunked at random, which means certain maps aren't valid, e.g. `Text.words <$> myRope` would give some unexpected results if chunk-splits happen in the middle of words, this is expected of course if you understand how Ropes work.

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 7 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Util/Rope.hs:97]: Or alternatively remove the size check from `snoc` and `cons` and keep the lower-level one.

File: parser-typechecker/src/Unison/Util/Rope.hs [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+  reverse = \case
+    One a -> One (reverse a)
+    Two sz l r -> Two sz (reverse r) (reverse l)
+    Empty -> Empty
+
+two :: Sized a => Rope a -> Rope a -> Rope a
+two r1 r2 = Two (size r1 + size r2) r1 r2
+
+threshold :: Int
+threshold = 32
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Rope.hs]: Any insight into where this came from? It seems quite small for things like ByteStrings and Text.

The [Rope for the Yi text editor](https://github.com/yi-editor/yi-rope/blob/master/src/Yi/Rope.hs#L186-L191) uses 1200 as their chunk size.

At any rate it'd be good to write down the rationale here so we can re-evaluate in the future (even if the current rationale is "it's a guess"

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 9 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs]: Yeah, though there's some sloppier notion of balance that even the result of a `take` or `drop` still satisfies. At the very least, if all the chunks were size 1, then a `take` or `drop` will still have depth `O(log N)` if the original tree also had logarithmic depth.

I was also thinking of splay tree style rebuilding in `take` / `drop`, which saves a log factor during purely sequential walking of the tree from one direction or the other, but that needs to be done carefully or you can end up with an overly unbalanced tree.

File: parser-typechecker/src/Unison/Util/Text.hs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  case T.decodeUtf8' (B.toByteString bs) of
+    Right t -> Right (fromText t)
+    Left e -> Left (show e)
+
+toUtf8 :: Text -> B.Bytes
+toUtf8 (Text t) = B.Bytes (R.map (B.chunkFromByteString . T.encodeUtf8 . chunkToText) t)
+
+fromText :: T.Text -> Text
+fromText s | T.null s = mempty
+fromText s = go (Text (R.one (chunk s)))
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Text.hs]: I think it may be more efficient to use `Text.chunksOf` to split the text into threshold-sized chunks, then you can build a balanced tree directly from that list.

Seems like this version has to `take` and `drop` over the same pieces of text many times whereas `chunksOf` could split into the correct size on the first go 😄

File: parser-typechecker/src/Unison/Util/Text.hs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  where
+  go t | n > threshold  = go (take (n `div` 2) t) <> go (drop (n `div` 2) t)
+       | otherwise      = t
+       where n = size t
+
+pack :: String -> Text
+pack = fromText . T.pack
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Text.hs:83]: Here as well, we have an opportunity to split the string into the proper size of chunks before `pack`ing it, IIRC splitting up text results in needing to copy the bytes of each split into completely new Text's; so splitting up the strings before packing them would hopefully avoid some of that overhead.

Not sure which functions we're most performance conscious about, so use your discretion on what's worth investigating 😄

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 12 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs]: It's a guess, but what you're going for is that a block copy of less than `threshold` is as fast or faster than creating a new subtree that points to the two blocks. [Scala's Vector](https://github.com/scala/scala/blob/v2.13.3/src/library/scala/collection/immutable/Vector.scala#L97-L99) uses 32 for this, as does Clojure's sequence type. 

If you're okay with worse performance on edits and want to optimize more for memory usage, then you can make the threshold higher. That makes sense for a text editor where the runtime just needs to be fast enough to be imperceptible.

Basically, by default, this data structure is meant to be as fast as possible for edits (for situations where it's being used as a text buffer), and if you want to later coalesce some chunks to improve memory usage, you could do that. (We could just expose a `Text.flatten` operation like we do for `Bytes`, and then you can achieve whatever chunk size you want - repeatedly `take 1200`, flatten, concat)

I'll add some color on this in comments though!

File: parser-typechecker/src/Unison/Util/Text.hs [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+import Prelude hiding (take,drop,replicate)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Unison.Util.Bytes as B
+import qualified Unison.Util.Rope as R
+
+-- Text type represented as a `Rope` of chunks
+newtype Text = Text (R.Rope Chunk) deriving (Eq,Ord,Semigroup,Monoid)
+
+data Chunk = Chunk {-# unpack #-} !Int {-# unpack #-} !T.Text
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Util/Text.hs:19]: Is this wrapper necessary? I'm curious if it would be feasible to add a size field to `One` on the Rope itself and just use that.
Are there other benefits to using Chunk that a `Rope Text` can't handle on its own?
Maybe `type Text = Rope Text` would suffice without the added indirection.

Or am I missing benefits from being able to `unpack` the text into here, which we can't do on a Rope due to polymorphism?

File: parser-typechecker/src/Unison/Util/Text.hs (hunk 13 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Text.hs]: Ah, good call!

File: parser-typechecker/src/Unison/Util/Text.hs (hunk 14 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Text.hs:83]: > IIRC splitting up text results in needing to copy the bytes of each split into completely new Text's; so splitting up the strings before packing them would hopefully avoid some of that overhead.

`Text` is represented with an underlying byte array and an offset and length into that array (hence why it exposes a [`copy`](https://hackage.haskell.org/package/text-1.2.4.1/docs/src/Data.Text.html#copy) function that copies just that slice to a fresh array). You can have multiple text values that share the same array, just with different offsets and lengths, and I would assume that like `Text.chunksOf` will not produce copies.

So doing the chunking on the string and then packing those would lose out on this sharing!

**Review by ChrisPenner — APPROVED** [review]
I asked a lot of questions, mostly performance related;

If performance is critical for these built-in types (which, since this is being implemented, it seems is the case), then I'd really love to have some benchmarks 😄 , they'd make me much more comfortable editing this code in the future 👍🏼

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 12 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Util/Rope.hs]: Awesome, thanks for the context! I learned something new today ✅

File: parser-typechecker/src/Unison/Util/Text.hs (hunk 14 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Util/Text.hs:83]: Ah, gotcha! That makes sense.

So if I'm understanding that correctly, the sharing would likely improve performance on creation, since we only allocate a single array and then swap some offsets around later;

But now I'm curious about memory residency. If I `pack` a very long string into a `Text`, then I `take` less than one threshold and throw away the rest, I'm still paying the memory cost of the whole array.

Whereas if we chunk, and then pack, each chunk has its own array, so if we `take` a few chunks, we only pay the memory residency of the text we actually use.

It seems `Text.append` doesn't check if we're appending two contiguous texts from the same array, so if you `take` perform any operation which involves an append we'll lose the sharing anyhow, so maybe my point is moot here.

Thanks for the clarification, really appreciate it 🙌🏼

Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Text.hs:83]: Yes, sharing of the underlying array is sort of a double-edged sword if you just `take 1` and then throw out the rest. You could have various operations pre-emptively copy, but somehow that doesn't seem like the right call.

It's probably a good idea to expose a `Text.copy` builtin.

File: parser-typechecker/src/Unison/Util/Text.hs (hunk 15 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Text.hs:19]: Yeah could add a size field to `One` (I had phrased it that way at one point) but then you have the size represented twice in the case where you have `Rope (Vector Word8)`.

I was hoping the `{-# unpack -#}` in `One` would give what I wanted, which is that the text and its size are unpacked into the constructor. If that won't happen then I'd probably just go back to the version that had the size in the `One`, and then get rid of the `Chunk` type for `Text` and just use `Rope Data.Text` directly. It's an extra `Int` for the `Bytes` case but that's probably fine.

**Comment by pchiusano** [comment]
Okay, thanks y'all, also thanks @dolio for filling in those functions!

I think what I'll do is add some more comments based on the discussion and either add a few basic benchmarks or create a ticket to track.

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 11 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs:38]: Yeah, I'll add a note. It's often okay to change the size of chunks. It might lead to a less size-balanced tree, but that's often fine. Like going from `Rope Text` to `Rope ByteString` by utf8 encoding each chunk is fine and doesn't need rebalancing

Mapping (say) 95% of the chunks to empty is where you want to rebalance.

File: parser-typechecker/src/Unison/Util/Bytes.hs (hunk 4 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Bytes.hs]: Went with `Chunk` as the name.

File: parser-typechecker/src/Unison/Util/Rope.hs (hunk 7 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Rope.hs:107]: Agree. No need to check in two places... will update.

**Comment by pchiusano** [comment]
Tracking benchmarking in https://github.com/unisonweb/unison/issues/2599

File: parser-typechecker/src/Unison/Util/Bytes.hs (hunk 3 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Util/Bytes.hs]: There aren't specific unit tests for these functions but those codepaths should get hit by a bunch of existing tests, yeah.
